cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala

```diff
@@ -237,43 +237,47 @@ extends ColumnStore with CassandraChunkSource with StrictLogging {
-      for (row <- rows) {
-        val partKeyRecord = PartitionKeysTable.rowToPartKeyRecord(row)
+    def copyRows(targetPartitionKeysTable: PartitionKeysTable, records: Set[PartKeyRecord], shard: Int) = {
+      val partKeys = records.map(partKeyRecord =>
         targetPartitionKeysTable.readPartKey(partKeyRecord.partKey) match {
-          case Some(targetPkr) =>
-            targetPartitionKeysTable.writePartKey(compareAndGet(partKeyRecord, targetPkr), diskTimeToLiveSeconds)
-          case None =>
-            targetPartitionKeysTable.writePartKey(partKeyRecord, diskTimeToLiveSeconds)
+          case Some(targetPkr) => pkRecordWithHash(compareAndGet(partKeyRecord, targetPkr))
+          case None => pkRecordWithHash(partKeyRecord)
         }
-      }
+      )
+      val updateHour = System.currentTimeMillis() / 1000 / 60 / 60
+      Await.result(
+        writePartKeys(targetDatasetRef, shard, Observable.fromIterable(partKeys), diskTimeToLiveSeconds, updateHour),
+        Duration.Inf // TODO - Use cassandra write timeout.
+      )
     }
 
     // for every split, scan PartitionKeysTable for all the shards.
-    for (split <- splits) {
-      for (shard <- 0 until numOfShards) {
-        val tokens = split.asInstanceOf[CassandraTokenRangeSplit].tokens
-        val srcPartKeysTable = getOrCreatePartitionKeysTable(datasetRef, shard)
-        val targetPartKeysTable = target.getOrCreatePartitionKeysTable(targetDatasetRef, shard)
-        // CQL does not support OR operator. So we need to query separately to get the timeSeries partitionKeys
-        // which were born or died during the data loss period (aka repair window).
-        val rowsByStartTime = srcPartKeysTable.scanRowsByStartTimeRangeNoAsync(tokens, repairStartTime, repairEndTime)
-        val rowsByEndTime = srcPartKeysTable.scanRowsByEndTimeRangeNoAsync(tokens, repairStartTime, repairEndTime)
-        // add to a Set to eliminate duplicate entries.
-        val rowSet = rowsByStartTime.++(rowsByEndTime)
-        copyRows(targetPartKeysTable, rowSet)
-      }
+    for (split <- splits; shard <- 0 until numOfShards) {
+      val tokens = split.asInstanceOf[CassandraTokenRangeSplit].tokens
+      val srcPartKeysTable = getOrCreatePartitionKeysTable(datasetRef, shard)
+      val targetPartKeysTable = target.getOrCreatePartitionKeysTable(targetDatasetRef, shard)
+      // CQL does not support OR operator. So we need to query separately to get the timeSeries partitionKeys
+      // which were born or died during the data loss period (aka repair window).
+      val rowsByStartTime = srcPartKeysTable.scanRowsByStartTimeRangeNoAsync(tokens, repairStartTime, repairEndTime)
+      val rowsByEndTime = srcPartKeysTable.scanRowsByEndTimeRangeNoAsync(tokens, repairStartTime, repairEndTime)
+      // add to a Set to eliminate duplicate entries.
+      val records = rowsByStartTime.++(rowsByEndTime).map(PartitionKeysTable.rowToPartKeyRecord)
```

Comment:
Try your best to use iterator instead of bringing everything into memory here.
There are times when contents of a single split may blow out the heap allocation. These things tend to happen at wrong times.

It may be ok to pay the cost of writing same part key twice into cassandra - that may be an acceptable trade off for the few number of such PKs that both start and end in the repair period. 

This will eliminate the need to implement equals on `PartitionKeyRecord` as well.